### DISCUSSION TITLE
`no-unsafe-finally` on generator functions

### DISCUSSION THREAD

**Discussion by davbrito** [eslint/eslint#17073] [Q&A]
I have a code like this. Where a generator function delegates the control flow to the caller for requesting inputs of data. Then the caller indicates to the generator that there is no more data by calling it's return method.

```js
function* concatMatching(pattern) {
  let value = "";
  try {
    while (true) {
      const string = yield;
      if (!pattern.test(string)) continue;
      value += string;
    }
  } finally {
    return value;
  }
}

const gen = concatMatching(/^[a-z]+$/);

gen.next();
gen.next("hello")
gen.next("World")
console.log(gen.return()); // { value: "hello", done: true }
```

Should this trigger an `no-unsafe-finally`? Is there a better or "more correct" way to achieve what I'm doing?

**Comment by sam3k** [comment]
Hey @davbrito, seems to me like it does trigger an [error](https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24qIGNvbmNhdE1hdGNoaW5nKHBhdHRlcm4pIHtcbiAgbGV0IHZhbHVlID0gXCJcIjtcbiAgdHJ5IHtcbiAgICB3aGlsZSAodHJ1ZSkge1xuICAgICAgY29uc3Qgc3RyaW5nID0geWllbGQ7XG4gICAgICBpZiAoIXBhdHRlcm4udGVzdChzdHJpbmcpKSBjb250aW51ZTtcbiAgICAgIHZhbHVlICs9IHN0cmluZztcbiAgICB9XG4gIH0gZmluYWxseSB7XG4gICAgcmV0dXJuIHZhbHVlO1xuICB9XG59XG5cbmNvbnN0IGdlbiA9IGNvbmNhdE1hdGNoaW5nKC9eW2Etel0rJC8pO1xuXG5nZW4ubmV4dCgpO1xuZ2VuLm5leHQoXCJoZWxsb1wiKVxuZ2VuLm5leHQoXCJXb3JsZFwiKVxuY29uc29sZS5sb2coZ2VuLnJldHVybigpKTsgLy8geyB2YWx1ZTogXCJoZWxsb1wiLCBkb25lOiB0cnVlIH0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjpbImVycm9yIl0sImZvci1kaXJlY3Rpb24iOlsiZXJyb3IiXSwiZ2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjpbImVycm9yIl0sIm5vLWNhc2UtZGVjbGFyYXRpb25zIjpbImVycm9yIl0sIm5vLWNsYXNzLWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb21wYXJlLW5lZy16ZXJvIjpbImVycm9yIl0sIm5vLWNvbmQtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1jb25zdGFudC1jb25kaXRpb24iOlsiZXJyb3IiXSwibm8tY29udHJvbC1yZWdleCI6WyJlcnJvciJdLCJuby1kZWJ1Z2dlciI6WyJlcnJvciJdLCJuby1kZWxldGUtdmFyIjpbImVycm9yIl0sIm5vLWR1cGUtYXJncyI6WyJlcnJvciJdLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOlsiZXJyb3IiXSwibm8tZHVwZS1lbHNlLWlmIjpbImVycm9yIl0sIm5vLWR1cGUta2V5cyI6WyJlcnJvciJdLCJuby1kdXBsaWNhdGUtY2FzZSI6WyJlcnJvciJdLCJuby1lbXB0eSI6WyJlcnJvciJdLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOlsiZXJyb3IiXSwibm8tZW1wdHktcGF0dGVybiI6WyJlcnJvciJdLCJuby1leC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjpbImVycm9yIl0sIm5vLWV4dHJhLXNlbWkiOlsiZXJyb3IiXSwibm8tZmFsbHRocm91Z2giOlsiZXJyb3IiXSwibm8tZnVuYy1hc3NpZ24iOlsiZXJyb3IiXSwibm8tZ2xvYmFsLWFzc2lnbiI6WyJlcnJvciJdLCJuby1pbXBvcnQtYXNzaWduIjpbImVycm9yIl0sIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6WyJlcnJvciJdLCJuby1pbnZhbGlkLXJlZ2V4cCI6WyJlcnJvciJdLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6WyJlcnJvciJdLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6WyJlcnJvciJdLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOlsiZXJyb3IiXSwibm8tbmV3LXN5bWJvbCI6WyJlcnJvciJdLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6WyJlcnJvciJdLCJuby1vYmotY2FsbHMiOlsiZXJyb3IiXSwibm8tb2N0YWwiOlsiZXJyb3IiXSwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjpbImVycm9yIl0sIm5vLXJlZGVjbGFyZSI6WyJlcnJvciJdLCJuby1yZWdleC1zcGFjZXMiOlsiZXJyb3IiXSwibm8tc2VsZi1hc3NpZ24iOlsiZXJyb3IiXSwibm8tc2V0dGVyLXJldHVybiI6WyJlcnJvciJdLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6WyJlcnJvciJdLCJuby1zcGFyc2UtYXJyYXlzIjpbImVycm9yIl0sIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjpbImVycm9yIl0sIm5vLXVuZGVmIjpbImVycm9yIl0sIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjpbImVycm9yIl0sIm5vLXVucmVhY2hhYmxlIjpbImVycm9yIl0sIm5vLXVuc2FmZS1maW5hbGx5IjpbImVycm9yIl0sIm5vLXVuc2FmZS1uZWdhdGlvbiI6WyJlcnJvciJdLCJuby11bnNhZmUtb3B0aW9uYWwtY2hhaW5pbmciOlsiZXJyb3IiXSwibm8tdW51c2VkLWxhYmVscyI6WyJlcnJvciJdLCJuby11bnVzZWQtdmFycyI6WyJlcnJvciJdLCJuby11c2VsZXNzLWJhY2tyZWZlcmVuY2UiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1jYXRjaCI6WyJlcnJvciJdLCJuby11c2VsZXNzLWVzY2FwZSI6WyJlcnJvciJdLCJuby13aXRoIjpbImVycm9yIl0sInJlcXVpcmUteWllbGQiOlsiZXJyb3IiXSwidXNlLWlzbmFuIjpbImVycm9yIl0sInZhbGlkLXR5cGVvZiI6WyJlcnJvciJdfSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==).

**Reply by davbrito** [reply]
Sorry if I didn't express myself correctly. What I mean is that I think that in that specific case, when there is no return inside the "try" block, eslint should not put that error

**Comment by fasttime** [comment]
It seems that the current behavior is intended, even if the documentation is not very clear and in fact suggests the opposite.

> JavaScript suspends the control flow statements of `try` and `catch` blocks until the execution of `finally` block finishes. So, when `return`, `throw`, `break`, or `continue` is used in `finally`, **control flow statements inside `try` and `catch` are overwritten, which is considered as unexpected behavior**.

But there are unit tests to ensure that an error is reported even when there are no control statements in `try` or `catch` blocks, for example:

https://github.com/eslint/eslint/blob/e5182b723ff82bb3b55c50c06d64626055414b31/tests/lib/rules/no-unsafe-finally.js#L86-L89

Normally, when there are no control flow statements to divert the code path in `try`/`catch` blocks, execution continues normally after the `try` statement, making a `finally` block unnecessary. For example, your function could be rewritten as:

```js
function* concatMatching(pattern) {
  let value = "";
  try {
    while (true) {
      const string = yield;
      if (!pattern.test(string)) continue;
      value += string;
    }
  } catch {
  }
  return value;
}
```

Granted, this is not better of shorter than your version, but it passes the rule 😄.
